src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs

```diff
@@ -74,53 +75,84 @@ public async Task<LSP.CompletionItem> HandleRequestAsync(LSP.CompletionItem comp
-            LSP.CompletionItem resolvedCompletionItem;
-            if (lspVSClientCapability)
+            if (completionItem is LSP.VSCompletionItem vsCompletionItem)
             {
-                resolvedCompletionItem = CloneVSCompletionItem(completionItem);
-                ((LSP.VSCompletionItem)resolvedCompletionItem).Description = new ClassifiedTextElement(description.TaggedParts
+                vsCompletionItem.Description = new ClassifiedTextElement(description.TaggedParts
                     .Select(tp => new ClassifiedTextRun(tp.Tag.ToClassificationTypeName(), tp.Text)));
             }
-            else
+
+            // We compute the TextEdit resolves for override and partial method completions here.
+            // Lazily resolving TextEdits is technically a violation of the LSP spec, but is
+            // currently supported by the VS client anyway. Once the VS client adheres to the spec,
+            // this logic will need to change and VS will need to provide official support for
+            // TextEdit resolution in some form.
+            if (completionItem.InsertText == null && completionItem.TextEdit == null)
             {
-                resolvedCompletionItem = RoslynCompletionItem.From(completionItem);
-                ((RoslynCompletionItem)resolvedCompletionItem).Description = description.TaggedParts.Select(
-                    tp => new RoslynTaggedText { Tag = tp.Tag, Text = tp.Text }).ToArray();
+                var snippetsSupported = context.ClientCapabilities.TextDocument?.Completion?.CompletionItem?.SnippetSupport ?? false;
+
+                completionItem.TextEdit = await GenerateTextEditAsync(
+                    document, completionService, selectedItem, snippetsSupported, cancellationToken).ConfigureAwait(false);
             }
 
-            resolvedCompletionItem.Detail = description.TaggedParts.GetFullText();
-            return resolvedCompletionItem;
+            completionItem.Detail = description.TaggedParts.GetFullText();
+            return completionItem;
         }
 
-        private static LSP.VSCompletionItem CloneVSCompletionItem(LSP.CompletionItem completionItem)
+        private static async Task<LSP.TextEdit> GenerateTextEditAsync(
+            Document document,
+            CompletionService completionService,
+            CompletionItem selectedItem,
+            bool snippetsSupported,
+            CancellationToken cancellationToken)
         {
-            return new LSP.VSCompletionItem
+            var documentText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var completionChange = await completionService.GetChangeAsync(
+                document, selectedItem, cancellationToken: cancellationToken).ConfigureAwait(false);
+            var completionChangeSpan = completionChange.TextChange.Span;
+            var newText = completionChange.TextChange.NewText;
+
```

Comment:
```suggestion
            var newText = completionChange.TextChange.NewText;
            Contract.ThrowIfNull(newText);
```

(and remove `?.`s below).